### PR TITLE
Fix: allow executing CopyCommand operations in parallel when it is set

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR fixes the CopyCommand execute in sequence issue observed in #5925 
Fixes: #5925

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 1]
```diff
@@ -342,10 +343,12 @@ public Answer sendTo(final Long dcId, final HypervisorType type, final Command c
                 }
                 Answer answer = null;
                 try {
-
-                    final long targetHostId = _hvGuruMgr.getGuruProcessedCommandTargetHost(host.getId(), cmd);
+                    HypervisorGuru hvGuru = _hvGuruMgr.getGuru(host.getHypervisorType());
+                    Pair<Boolean, Long> result = hvGuru.getCommandHostDelegation(host.getId(), cmd);
+                    final long targetHostId = result.first() ? result.second() : host.getId();
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:348]: isn't this bit responsibility of the `HypervisorGuruManager` ?

**Comment by nvazquez** [comment]
Thanks @DaanHoogland - refactored
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by Pearl1594 — APPROVED** [review]
LGTM. Verified fix - initiated volume migration and start a VM deployment on the same host handling volume migration and the 2 operations ran concurrently. This was done with global setting vmware.create.full.clone set to false. On setting vmware.create.full.clone to true, the 2 commands are handled sequentially by the host.

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 2]
```diff
@@ -342,10 +342,11 @@ public Answer sendTo(final Long dcId, final HypervisorType type, final Command c
                 }
                 Answer answer = null;
                 try {
-
-                    final long targetHostId = _hvGuruMgr.getGuruProcessedCommandTargetHost(host.getId(), cmd);
+                    final long targetHostId = _hvGuruMgr.getGuruProcessedCommandTargetHost(host.getId(), cmd, host.getHypervisorType());
                     answer = easySend(targetHostId, cmd);
                 } catch (final Exception e) {
+                    s_logger.error("Error sending command " + cmd.getClass().getName() + " to host: " +
+                            host.getUuid(), e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:349]: please do not include the stacktrace on error level, on debug level is good enough. Most operators will not have a use for it and only be alarmed. 
```suggestion
                    s_logger.error(String.format("Error sending command %s to host %s, due to %s", cmd.getClass().getName(),
                            host.getUuid(), e.getLocalizedMessage()));
                    if (s_logger.isDebugEnabled()) {
                        s_logger.debug(String.format("Error sending command %s" + " to host %s",
                            cmd.getClass().getName(),
                            host.getUuid(), e));
                    }
```

Comment (nvazquez) [reply engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java]: Done, thanks

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Pull request merged**